base-files: rework system.sh functions

mcprat opened the pull request:
all functions in system.sh (mostly for MAC) reworked and cleaned up
with a focus on:

 - readable code
 - function reuse
 - exposing parameter options
 - stronger filtering
 - strict syntax

Summary:

 - prefer head/tail/builtins over hexdump/dd/sed (maybe switch to od in future)
 - mtd_get_mac_ascii	fully rewritten to handle corner cases, string filtering
 - macaddr_geteui	converted to new function macaddr_octet with options
 - macaddr_canonicalize	handle all possible sets of delimiters between octets
 - all functions	syntax cleaned up, checked with shellcheck

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
base-files: cleanup system.sh functions

all functions in system.sh (mostly for MAC) cleaned up
with a focus on:

 - readable code
 - function reuse
 - exposing more parameter options
 - wrapper functions use $@ for input
 - strict syntax
 - prefer shell methods over busybox tools when possible
 - checked with shellcheck

the setbit and unsetbit functions now also canonicalize
the canonicalize function now also trims hex to length

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
reference conversation #9468 #4199

mpratt14 commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -219,17 +259,26 @@ macaddr_canonicalize() {
+	[ "${#canon}" -eq 17 ] || return
 
-	printf "%02x:%02x:%02x:%02x:%02x:%02x" 0x${canon// / 0x} 2>/dev/null
+	printf '%02x:%02x:%02x:%02x:%02x:%02x' "0x${canon// / 0x}" 2>/dev/null
```
something I just realized, double quotes breaks printf...

mpratt14 commented:
probably the final version now

mpratt14 commented:
ok it works now

mpratt14 commented:
there was a typo

mpratt14 commented:
ping @DragonBluep if you would like to test

obsy commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -42,48 +49,67 @@ get_mac_label_json() {
-	mac_dirty=$(strings "$part" | sed -n 's/^'"$key"'=//p')
+	value=$(strings "$part" |
+		while read a; do
+			[[ "$a" == *"$key"* ]] && echo -n "$a" && break
```
NACK. Bashism. Don't use [[ and ==, see many other openwrt commits.

jow- commented:
On OpenWrt busybox ash, `head` and `tail` are not builtins but external programs.

jow- commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -1,31 +1,36 @@
 
-	hexdump -v -n 6 -s $offset -e '5/1 "%02x:" 1/1 "%02x"' $path 2>/dev/null
+	mac=$(head -c $((offset + length)) "$path" | tail -c "$length" |
+		hexdump -e '6/1 "%02x"' 2>/dev/null)
```
This replaces a single fork/exec with one subshell and two fork/exec to invoke external `head` and `tail` instead of utilizing hexdump's builtin skipping.

jow- commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -1,31 +1,36 @@
+	[ -e "$basepath/$macdevice/mac-address" ] &&
+		macaddr=$(get_mac_binary "$basepath/$macdevice/mac-address" 2>/dev/null) ||
+		macaddr=$(get_mac_binary "$basepath/$macdevice/local-mac-address" 2>/dev/null)
 
```
Strictly speaking this changes logic. If `get_mac_binary()` would result an empty result in the old code (e.g. while reading a zero-length file), the fallback logic would be used. Now it decides solely on the presence of the file (and introduces a redundant stat syscall instead of a simple string length check)

jow- commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -42,48 +49,67 @@ get_mac_label_json() {
+	local macaddr
 
-	[ -n "$macaddr" ] || macaddr=$(get_mac_label_json)
+	[ -n "$(get_mac_label_dt)" ] &&
```
This invokes `get_mac_label_dt()` twice in the worst case. The result should be cached in an intermediate variable.

jow- commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -42,48 +49,67 @@ get_mac_label_json() {
+	value="${value//$operators/}"
+
+	# if quoted
+	[[ "$value" != "${value//$blockchar/}" ]] &&
```
Please use single bracket notation (`[ ... ]`)

jow- commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -42,48 +49,67 @@ get_mac_label_json() {
+	suffix="${value#*${delimiter}}"
+
+	# if no delimiter
+	[[ "$prefix" == "$suffix" ]] &&
```
Please use single bracket notation (`[ ... ]`)

jow- commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -42,48 +49,67 @@ get_mac_label_json() {
 
 mtd_get_mac_text() {
-	local mtdname=$1
+	local mtdname="$1"
```
Quoting on assignment is redundant (I know it's being done elsewhere in the file, but I'd rather fix those other cases)

mpratt14 replied:
most examples I read have variables quoted, and some say its good practice

the only context for no quotes being:

> In contexts where we are absolutely sure that there are no word splitting or globbing issues or we do want word splitting and globbing"

I know these functions dont expect a space or wildcard `*` but if some random character is thrown in we don't want it to be interpreted as an operator and do something unexpected before returning

jow- commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -96,10 +122,9 @@ mtd_get_mac_text() {
 	fi
 
-	mac_dirty=$(dd if="$part" bs=1 skip="$offset" count=17 2>/dev/null)
+	mac=$(head -c $((offset + length)) "$part" | tail -c "$length")
```
Introduces an additional fork/exec instead of utilizing `dd`'s builtin skipping.

jow- commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -114,24 +139,21 @@ mtd_get_mac_binary() {
-			echo $((0x$size))
-			break
-		fi
+		[[ "$name" == *"$mtdname"* ]] && echo -n $((0x${size})) && break
```
Bashsim. Use 

```bash
case "$name" in
  *"$mtdname"*) echo -n $((0x${size})) && break ;;
esac
```

jow- commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -145,71 +167,85 @@ mmc_get_mac_binary() {
-	local mac=$1
-
-	echo -ne \\x${mac//:/\\x}
+	echo -n -e "\\x${1//$delimiter/\\x}"
```
Consider using printf

mpratt14 commented:
thanks for the quick review, I addressed almost all of it. I also found a trick for some of them to accept stdin as input.

`dd` vs `head | tail` is an interesting case though...

I messed around with measuring performance between them with `time -v` and something stands out

`head | tail`

```
...
        Voluntary context switches: 1
        Involuntary context switches: 2
...
        Voluntary context switches: 3
        Involuntary context switches: 0
...

```

`dd`

```
...
        Voluntary context switches: 1
        Involuntary context switches: 35
...
```

someone could correct me on the meaning or if it really means anything at all, but it seems like head and tail are more "efficient" with the cpu usage, whether or not there is more or less cpu usage (time is 0.01s more)

obsy commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -1,105 +1,122 @@
+	suffix="${value#*${delimiter}}"
+
+	# if no delimiter
+	[ "$prefix" == "$suffix" ] &&
```
Single =

mans0n commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -1,105 +1,122 @@
 	fi
 
-	mac_dirty=$(strings "$part" | sed -n 's/^'"$key"'=//p')
+	value=$(strings "$part" |
+		while read vars; do
+			case "$vars" in
+				*"$key"*) echo -n "$vars" && break ;;
+			esac
+		done)
```
This might result a false positive. https://github.com/openwrt/openwrt/pull/9468#issuecomment-1066621738 https://github.com/openwrt/openwrt/pull/9468#issuecomment-1066839862

If we're looking for "addr" and found "Caddr", then we cannot know whether there is a valid "Caddr" key, or "addr" key is prefixed by the CRC. IMO dropping the entire CRC bytes is the only way to solve the problem robustly.

mpratt14 replied:
ok so after some thinking

 - added automatic and manual way to deal with CRC bytes, or something similar
 - globbing on both sides of "key" is a backup, globbing after only takes priority

mans0n replied:
It is still based on so many assumptions and conditions. It might work for most cases but it opens more holes to corner cases. (e.g. `"$key"*` will also take `ethaddr1` even the input key is `ethaddr`) I hope there's a simpler approach to this...

mpratt14 replied:
good point, however we are starting to get into cases that are more and more unlikely. 

so let's just have the break condition an exact match, and glob otherwise?

mans0n replied:
Why don't we just drop the CRC bytes and use the strict matching as before? (no globbing)
IMHO some of your changes seem to be unnecessary complicated, personally I'd prefer to keep the change as minimal, except for the shellcheck fix (thanks for taking care of it BTW).

mpratt14 replied:
when you last commented I had already placed a way to skip the CRC

now the function also prefers a non-globbing match over a globbing match, in other words globbing is a fallback

DragonBluep replied:
> 

Sorry for the lateness, but not all environment variables start at the beginning of the mtd partition. For example, the RAISECOM MSG1500 you mentioned, its environment variables are divided into two parts, one part starts from u-boot-env@0x0, and the other part starts from u-boot-env@0x8000. Your approach is good, I think for these more special devices, we can use "fw_printenv" or "fw_printsys" directly, as @mans0n once said.

mans0n replied:
BTW have you really tested it on device? Try calling `mtd_get_mac_ascii` with `set -x` and it will be evident what is really going on.

mpratt14 replied:
I have been testing during writing and after pushing intermittently, I've been doing more serious testing recently compared to a few days ago

mpratt14 replied:
@DragonBluep that second environment @ 0x8000 has its own CRC as well? is it a different mtd partition too?

mans0n replied:
Okay, I'm sorry for being harsh.
What I wanted to report:
```
root@OpenWrt:/tmp# mtd_get_mac_ascii config ethaddr
root@OpenWrt:/tmp# set -x
root@OpenWrt:/tmp# mtd_get_mac_ascii config ethaddr
+ mtd_get_mac_ascii config ethaddr
...
+ keys='bootcmd=tftp
ethaddr=xx:xx:xx:xx:xx:xx
...
stdout=serial'
+ mac='ethaddr=xx:xx:xx:xx:xx:xx'
+ continue
+ echo -n '=xx:xx:xx:xx:xx:xx'
+ macaddr_canonicalize
+ local 'mac='
+ local 'canon='
+ echo -n
+ tr -d '"'
+ mac=
+ '[' 0 -gt 17 ]
+ '[' -n  ]
+ '[' 0 -ne 17 ]
+ return
root@OpenWrt:/tmp# 
```

> that second environment @ 0x8000 has its own CRC as well?

It seems so. https://github.com/openwrt/openwrt/pull/9468#issuecomment-1066308396

mpratt14 replied:
so I did some testing with `set -x` (thanks for that tip)

yeah I see the mistake I made, true "exact match" of "key" will never happen because we are looking for "key...value"

however, canonicalize function would filter the "=", it could be that when you tried the ascii function it was with the original canonicalize, or without the char classes I made....or the char class was broken...

mpratt14 commented:
the "read stdin" trick stopped working the way i thought it was

might have to get rid of it

mpratt14 commented:
the functions I had changed to be able to read stdin are now working with parameters too

I reworked the canonicalize function again, using printf again and a new function for splitting instead of bash-like `${var:offset:length}`

mpratt14 commented:
setbit and unsetbit functions changed to start at index 0

mans0n commented:
I suggest you to split your system.sh patch into multiple pieces, it's really hard (at least for me) to review.
Personally I don't really like to change the functions in incompatible manner, and if we do it should be well-justified. Because I regard it as some kind of system tool library, and some downstream projects might be also relying on this.

mans0n commented on package/base-files/files/lib/functions/system.sh:
```diff
@@ -145,91 +156,131 @@ mmc_get_mac_binary() {
-	local mac=$1
-	local bit=${2:-0}
+	local hex="${1//$delimiter/}"
+	local bit="${2:-0}"
```
The reason why I set this to 0 as default in dc61e3b7ffdc was because I wanted to make sure the function returns nothing if `bit` argument is not defined. If you're going to make it start from 0 (which I don't like), then I think the default value should be changed too. (why prefer certain bit over others?)

mpratt14 commented:
@mans0n 
> I suggest you to split your system.sh patch into multiple pieces

yeah I agree, although I dont know *how much* i can do so

I just pushed what I had changed recently so I don't lose track of the changes (character classes were incorrect, and a few small things)



> I don't really like to change the functions in incompatible manner

yes they should all be backwards compatible.... I was just thinking that we usually start counting bits at 0, and otherwise theres a small confusion (other than the reverse order)

should I just put it back or remove the default value?

Thinking more on that, the stdin trick is not the best way (I was going to propose something later that replaces it), so it's not worth the trouble of adding it here.

mcprat committed:
base-files: add extra ASCII character classes

the POSIX defined character class [:punct:] covers
all of the characters in these classes, but sometimes
it may be desired to treat them as separate catagories
as they have different roles in scripting.

tested by creating a string of all printable ASCII
and then filtering out each class.

  root@OpenWrt:~# string=$(awk 'BEGIN{for(i=32;i<127;i++)printf "%c",i; print}')
  root@OpenWrt:~# string=${string//$delimiter/}
  ...
  root@OpenWrt:~# echo $string

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mans0n commented:
> I was just thinking that we usually start counting bits at 0, and otherwise theres a small confusion (other than the reverse order)

Yeah, I agree that starting with 1 is not the best option. (I don't remember why I did it in that way) It's just that I'm not fond of changing the current behavior unless we have a good reference for the bit counting rule for MAC address.

mcprat committed:
base-files: replace function macaddr_geteui with macaddr_octet

the function macaddr_geteui is currently unused.
it was made to grab only the last 3 octets of a MAC

replace it with a universal, POSIX-friendly function macaddr_octet
which can return any subset of octets as a single string of hex chars,
with or without a delimiter string to separate the octets.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
base-files: add function macaddr_split

replace the long case statement for splitting multi-octet strings
with a POSIX-friendly function that handles those cases and more.

in odd numbered strings, the first character is split first,
and then the rest in groups of 2 characters.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
base-files: rework function mtd_get_part_size

in a single bracket expression, globbing is not available,
but in case statements it is

also remove useless extra variable, prefer printf over echo

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
base-files: rework function macaddr_canonicalize

 - use character classes for filtering ASCII / hex
 - filter all non-hex or non-delimiter, instead of testing for
 - return when hex not longer than 1 for each octet
 - trim when given long hex string
 - append "0x" before printing
 - simple word count to check num of octets is sane, otherwise return

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
base-files: rework function mtd_get_mac_ascii

there is an issue where the CRC of u-boot-env
can occasionally have printable ascii chars,
making either the matching or parsing of the MAC
more difficult with sed when the desired key
is right after the CRC.

to address this and improve otherwise,
replace sed with shell tools and...

 - add optional command to deal with the partition manually
 - filter CRC with tail for u-boot-env partitions
 - only parse strings with length 12 or more
 - use globbing fallback if exact match fails

Reported-by: Shiji Yang <yangshiji66@qq.com>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
I split up all the changes that have more weight than the rest

the last commit might still have an issue, and I need to redo testing

after that, I would reorder the first to be the last


> a good reference for the bit counting rule for MAC address

wikipedia 😆 

I reverted the MAC bit index change for now 👍🏼

mpratt14 commented:
what I actually thought of was how bits in a register in a SOC datasheet are indexed

to be honest it seems to not matter anyway

it seems that the real standard is that octets are numbered from left to right from 0
but the bits in each octet are numbered from right to left in each octet from 0

too complicated for our use case

https://standards.ieee.org/wp-content/uploads/import/documents/tutorials/eui.pdf
https://standards.ieee.org/wp-content/uploads/import/documents/tutorials/macgrp.pdf

second document says "Individual / Group address bit (first bit transmitted on the medium)"
the way it is displayed that means its the LSB transmitted first, but in hex printed as MSB first

there's no way we are going to count like

`7  6  5  4  3  2  1  0  15 14 13 12 11 10  9  8  23 .....`

so I don't care about that anymore, its up to you

mans0n commented:
Thanks for the documents. Actually I haven't checked other than the wikipedia page...
It might be more intuitive to count the bits (in the order you described) as "Xth bytes and Yth bits" using two arguments, but if we are not going to take that way I think we can just leave it as is.

mpratt14 commented:
@mans0n I am thinking to just remove `macaddr_geteui` instead of improving it with `macaddr_octet`. a subset of octets can be returned with shell variable expansions (like how `macaddr_add` is now), and while the function can be used in other mac functions, there would not be any real improvement...

for example, the original use case for 6170c46b477d4953f91b99e805a276de444913cf
(note that it was reverted, except for the function 22b8a6263d72bb1911c5841ef90ebdf452b53905 )

```
root@OpenWrt:~# mac="aa:bb:cc:dd:ee:ff"
root@OpenWrt:~# hostname="Openwrt-$(mac=${mac#*:*:*:}; echo ${mac//$delimiter/})"
root@OpenWrt:~# echo $hostname
Openwrt-ddeeff
```

anything that can be done in 1 line shouldn't be a function...

However, I also think of someday removing the bash compatible operations from ash (a compile option) like `${var///}` and `${var:#:#}`, if we ever do that, then we would need a function with loops to remove the delimiters.

or, I can keep the new version, `macaddr_octet`, and simply not use it anywhere

what do you think?

mcprat committed:
ramips: RAISECOM MSG1500 skip first var environment and CRC

mtd_get_mac_ascii can now accept custom commands for handling
the partition that is named in the first parameter, for unusual cases.

this device has 2 variable environments,
the one storing MAC addresses at 0x8000 of the partition,
length of CRC is 4 bytes (u-boot)

when the CRC has printable characters,
matching and parsing the MAC can fail,
so skip the first 0x8004 bytes (+ 1 because of tail syntax)

Tested-by: Shiji Yang <yangshiji66@qq.com>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
ok so now....

 - added a specific commit for the RAISECOM device, to skip to byte `0x8005`
 - improved filtering of canonicalize function, no longer needs to check for non-hex chars
 - improved matching in ascii function to include partitions named `u-boot-env2` and so on (realtek devices have this)

@DragonBluep at this point you can test all this on your device to finally get a mac this way. I have no experience with nand devices though...I am pretty sure when you `cat` or `tail` the partition, that OOB / ECC data is not included, correct? for example you can try something like `cat /dev/mtdX | hexdump -C` and the variables still start at `0x8000`?

DragonBluep commented:
Thanks for taking care of this device. After my test, with the help of your patch, it works fine now.

> @DragonBluep at this point you can test all this on your device to finally get a mac this way. I have no experience with nand devices though...I am pretty sure when you `cat` or `tail` the partition, that OOB / ECC data is not included, correct? for example you can try something like `cat /dev/mtdX | hexdump -C` and the variables still start at `0x8000`?

Yes, partitions do not contain oob data, the "mtd" hardware abstraction layer handles and masks oob automatically.

mpratt14 commented:
added your tested-by

mpratt14 commented:
being very picky now, `|` is now in the operators class instead of delimiters

mpratt14 commented:
I think this is the last little update, final version 👍🏼

Djfe commented:
What kept this from being merged into master for release 23? 😅

mpratt14 commented:
it's not too late, I have to rebase

mcprat committed:
base-files: rework function mtd_get_mac_encrypted_arcadyan

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
base-files: rework function mtd_get_mac_encrypted_deco

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
ready for review again

@csharper2005 @mar-kolya would you be available to test changes to the functions that you added?

csharper2005 commented:
Hi! mtd_get_mac_encrypted_arcadyan() works fine. I checked it on official SNAPSHOT build + system.sh and functions.sh from your commit.

On 4/10/2023 8:37 PM, Michael Pratt wrote:
> ready for review again
> 
> @csharper2005 @mar-kolya would you be available to test changes to the functions that you added?
> 

-- 
Best regards,
Mikhail
